modules/azuread-application/main.tf

```diff
@@ -8,34 +8,45 @@ terraform {
   required_providers {
     azuread = {
       source  = "hashicorp/azuread"
-      version = "~> 1.3"
+      version = "~> 2.22"
     }
   }
 }
 
 resource "azuread_application" "app" {
-  display_name               = var.name
-  homepage                   = var.homepage
-  identifier_uris            = var.identifier_uris
-  reply_urls                 = var.reply_urls
-  logout_url                 = var.logout_url
-  available_to_other_tenants = var.available_to_other_tenants
-  public_client              = var.public_client
-  oauth2_allow_implicit_flow = var.oauth2_allow_implicit_flow
-  group_membership_claims    = var.group_membership_claims
-  owners                     = var.owners
-
-  dynamic "oauth2_permissions" {
-    for_each = var.oauth2_permissions
-
-    content {
-      admin_consent_description  = oauth2_permissions.value.admin_consent_description
-      admin_consent_display_name = oauth2_permissions.value.admin_consent_display_name
-      value                      = oauth2_permissions.value.value
-      type                       = oauth2_permissions.value.type
-      is_enabled                 = oauth2_permissions.value.is_enabled
-      user_consent_description   = oauth2_permissions.value.user_consent_description
-      user_consent_display_name  = oauth2_permissions.value.user_consent_display_name
+  display_name    = var.display_name
+  identifier_uris = var.identifier_uris
+
+  group_membership_claims = var.group_membership_claims
+  owners                  = var.owners
+
+  sign_in_audience               = var.sign_in_audience
+  fallback_public_client_enabled = var.fallback_public_client_enabled
+
+  web {
+    homepage_url  = var.homepage_url
+    logout_url    = var.logout_url
+    redirect_uris = var.redirect_uris
+
+    implicit_grant {
+      access_token_issuance_enabled = var.oauth2_implicit_flow_allow_access_token
+    }
+  }
+
+  api {
+    dynamic "oauth2_permission_scope" {
+      for_each = var.oauth2_permission_scopes
+
+      content {
+        admin_consent_description  = oauth2_permission_scope.value.admin_consent_description
+        admin_consent_display_name = oauth2_permission_scope.value.admin_consent_display_name
+        value                      = oauth2_permission_scope.value.value
+        type                       = oauth2_permission_scope.value.type
+        enabled                    = oauth2_permission_scope.value.enabled
+        user_consent_description   = oauth2_permission_scope.value.user_consent_description
+        user_consent_display_name  = oauth2_permission_scope.value.user_consent_display_name
+        id                         = uuid()
```

Comment:
Should this use uuidv5? It looks like it will generate a new uuid() on every apply.

```suggestion
        id                         = uuidv5("${oauth2_permission_scope.value.value}${oauth2_permission_scope.value.admin_consent_display_name}${oauth2_permission_scope.value.user_consent_display_name}${tostring(oauth2_permission_scope.value.enabled)}")
```

Their docs suggest using [random_uuid](https://registry.terraform.io/providers/hashicorp/random/latest/docs/resources/uuid) 
https://registry.terraform.io/providers/hashicorp/azuread/1.6.0/docs/resources/application#enabled

Their example:
https://github.com/hashicorp/terraform-provider-azuread/blob/9b133afd54e56bd601228b0a9243040524b0a2ec/examples/application/main.tf#L1-L13

If we were to go by their example then we would need to introduce the resource and pull the correct id, which would end up being something like:

```
resource "random_uuid" "oauth2_permission_scope_id" {
  for_each = var.oauth2_permission_scopes
}

...
...
    dynamic "oauth2_permission_scope" {
      for_each = var.oauth2_permission_scopes

      content {
        admin_consent_description  = oauth2_permission_scope.value.admin_consent_description
        admin_consent_display_name = oauth2_permission_scope.value.admin_consent_display_name
        value                      = oauth2_permission_scope.value.value
        type                       = oauth2_permission_scope.value.type
        enabled                    = oauth2_permission_scope.value.enabled
        user_consent_description   = oauth2_permission_scope.value.user_consent_description
        user_consent_display_name  = oauth2_permission_scope.value.user_consent_display_name
        id                         = random_uuid.oauth2_permission_scope_id[oauth2_permission_scope.key].result
```

Reply:
As per our conversation we would probably need to change `var.oauth2_permission_scopes` from `set(object())` to `map(object())` in order to have unique keys. When we have that changed to a `map()` then we can use `for_each` on a `resource "random_uuid"`.

Don't forget to add the provider to the `terraform{}` map:
```hcl
    random = {
      source = "hashicorp/random"
      version = "3.2.0"
    }
```